developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,349 @@
+messages, which are never sent to the DSP, so the firmware doesn't need to
+support them. For the most part IPCs from guests are forwarded without change by
+the host driver to the DSP, only some of them are intercepted on the host. Some
+such IPC messages trigger additional actions on the host before being forwarded
+to the DSP, others aren't sent to the DSP at all. See examples below for more
+details.
+
+2.1. Communication channels
+---------------------------
+
+The goal of the SOF virtualisation is to implement low latency audio
+virtualisation, which means avoiding copying of audio data and stream position
+updates. In this case only one VirtQueue would be required for IPC message
+exchange. However, the initial version also transfers data and stream position
+updates via VirtQueues. That means, that currently the SOF uses 3 VirtQueues:
+for IPC, audio data, and stream position updates. It is planned to support both
+the copying and the zero-copy methods in the future. The copying method will be
+used either where the zero-copy method is unavailable or for debugging.
+
+The low latency requirement also forces us to only support the vhost mode, there
+is no plan to implement SOF VirtIO host support in the user-space.
+
+   ..	figure::  images/data-flow.png
+	:align: center
+	:alt: SOF VirtIO Architecture
+	:width: 800px
+
+	`SOF virtualisation architecture`
+
+2.2. Topology
+-------------
+
+When running natively, the SOF driver uses the ALSA topology API to configure
+audio pipelines on the DSP. The SOF driver reads a platform-specific topology
+file from the filesystem, parses it and sends all the components from it to the
+DSP firmware for their instantiation, configuration and enablement.
+
+When configured to support virtualised audio, the host topology additionally
+provides connection points for guest audio pipelines. Guest playback pipelines
+might for example connect to mixer and capture pipelines to demultiplexer
+components. This way both the host and possibly multiple guests can share parts
+of the audio topology. It is also possible to dedicate certain audio pipelines
+to specific guests only. In such cases guest pipeline fragments can be connected
+to other component types, e.g. to a volume control. We'll use the term
+"connection points" throughout this document to refer to these locations in the
+host audio topology.
+
+The ALSA / ASoC / DAPM subsystems on the host are unaware of any guests, but
+they should be aware of any streaming, taking place on host parts of VM
+pipelines for configuration and power management. That means, e.g. if a guest
+initialises playback, the ASoC subsystem on the host should be aware, that the
+pipeline, joining the respective connection point and the used codec is
+currently performing audio playback. To achieve this, for each such guest
+connection point we add a virtual PCM as a widget of type "aif_in" or "aif_out"
+for playback and capture respectively, and a virtual 0-size buffer. They are
```

Comment:
@plbossart I think it's ALSA topology tools, that mandate one side of every topology connection to be a buffer. So, on the host in principle I just need
```
[ fake / virtual PCM ]
[ connection point, e.g. a mixer]
...
[codec]
```
where the fake / virtual PCM is for DPCM route updating on the host and is not sent to the DSP. But such a pipeline is invalid, there *must* be a buffer between that PcM and the mixer. And since we don't really need that buffer and it isn't sent to the DSP, we make it 0-byte large.

Reply:
It's worth adding that this is an ALSA topology restriction, I thought it was a virtio one.